---
title: "Southern Ontario Wastewater Analysis Report"
author: "Art Poon, Devan Becker, Gopi Gugan and Erin Brintnell"
date: "`r Sys.Date()`"
always_allow_html: true
output: 
  tufte::tufte_handout: default
  pdf_document:
    toc: false
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
---

```{r setup, include=FALSE}
library("here") # All paths relative to the directory with a .git folder
    # here("scripts", "plots.R") will always find plots.R,
    # no matter where this rmd's working directory is
library(flexdashboard)
library(dplyr)
library(tidyr)
library(knitr)
library(ggplot2)

omicron <- read.csv(here("reports", "omicron-mutations.csv")) %>%
  select(type, pos, alt, mutation) %>%
  #mutate(pos = pos-1) %>% 
  unite(label, type, pos, alt, sep = "|")
omicron_v_pango = read.csv(here("reports", "compare-devan-art.csv")) %>%
  mutate(pango_freq = pango_matches/220242,
    downstream = freq > 0.95 & pango_freq < 0.05) %>% 
  select(freq, pango_freq, label, downstream) %>%
  left_join(omicron, by = "label") %>% 
  mutate(mutation = ifelse(mutation == "None", label, mutation))
```


<!-- ## Sourcing existing lineages -->

<!-- ```{r sourceascript, eval=FALSE} -->
<!-- source(here("scripts", "plot.R")) # won't run - files don't exist on my machine -->
<!-- ``` -->

# Background

* On November 23, 2021, a request for a new lineage designation was submitted to the CoV-lineages PANGO designation GitHub site, describing a new sub-lineage of B.1.1 sampled in Botswana, Hong Kong and South Africa with an excessive number of mutations.

* The request was reviewed and accepted on November 24, designating the new lineage as B.1.1.529, designated "Omicron".

* On November 27, the CoV-lineages team uploaded a `constellation' file, describing a set of mutations for classifying genomes into the newly defined lineage.

## Objectives

1. identify common mutations in lineage B.1.1.529 (Omicron) relative to the SARS-CoV-2 reference genome sequence (WH1);

2. determine which of these common mutations are unique to B.1.1.529, in comparison to all other defined lineages;

3. retrospectively screen all available wastewater sample data sets for the presence of these mutations.


# Analysis of Omicron Mutations


Sequence data for available Omicron genomes was obtained (n=77). Gaps in Omicron sequence data were removed and data was processed to generate a list of mutations, indels and sections of low sequence coverage for each genome. 

The frequency of selected mutations in the Omicron genomes (n=77) and Pango lineages (n=220242) was calculated by counting the number of occurrences of each mutation. Frequency of mutations in Omicron and Pango lineages were compared and mutations present in >95\% of Omicron genomes and <5\% of all other Pango lineages were selected for further downstream analysis of wastewater data. 

There were `r nrow(omicron_v_pango)` mutations present in the Omicron genomes (n=77). Many mutations were present in other Pango lineages (**Figure 1**). Nonetheless, `r sum(omicron_v_pango$downstream)` mutations were present in >95\% of Omicron genomes and <5\% of all other Pango lineages (**Table 1**) which were used for downstream analyses.


```{r, echo = FALSE,fig.height=4.5}
ggplot(data = omicron_v_pango, 
      mapping = aes(x = freq, y = pango_freq + 0.000005, 
        label = mutation, colour = downstream)) + 
    geom_point() + 
    labs(x = "Frequency in Omicron", y = "Frequency in Pango Lineages",
      colour = "UsedIn Downstream Analysis") +
    scale_y_log10() + 
    geom_jitter() + 
    theme(legend.position="bottom", 
      panel.background = element_blank(),  
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_line(colour = "grey"), 
      axis.line = element_line(colour = "black"), 
      text = element_text(size = 8)) +
    scale_color_manual(values=c("#999999", "#E69F00")) +
    annotate("line", x = c(0.95, 0.95), y = c(0, 1), 
      col = "#E69F00", linetype = 2) +
    annotate("line", x = c(0, 1), y = c(0.05, 0.05), 
      col = "#E69F00", linetype = 2)

```

**Figure 1** Illustration of `r nrow(omicron_v_pango)` mutations found throughout 77 Omnicron genomes. Mutations used in downstream analysis were present in >95\% of Omicron genomes and <5\% of all other Pango lineages. These numbers were chosen arbitrarily and will be explored in further studies.



\pagebreak
**Table 1** Mutations selected for downstream analysis in wastewater samples.



```{r, echo=FALSE}
omicron_v_pango %>%
  filter(downstream) %>% 
  select(mutation, freq, pango_freq) %>%
  arrange(pango_freq) %>% 
  kable(col.names = c("Mutation", "Frequency in Omicron", "Frequency in Pango lineages"))
```



# Analysis of Omicron in Ontario Wastewater Samples

Coverage of genomes located at the positions of selected Omicron mutations was collected from sequence coverage files. Mapped wastewater reads were then probed for selected Omicron mutations. Omicron mutations present in wastewater samples was then scaled by coverage at the site.

Many wastewater samples had very low coverage (**Figure 2**), however there was evidence of some Omicron mutations in wastewater samples when scaled by coverage (**Figure 3**). Of note, aa:S:T547K, aa:S:G446S, aa:S:G339D, aa:orf1a:K856R were all present in >1% of genomes from samples processed by laboratories at the University of Guelph. Additionally, there was no evidence of any Omnicron mutations in samples processed at Western University of laboratories.   




\clearpage
\includegraphics[width=0.9\textwidth]{coverage.pdf}

**Figure 2** Coverage of wastewater samples across the SARS-CoV-2 genome. Low coverage means that we cannot tell if the mutations were not present or were simply not sequenced.



\clearpage

```{r, echo=FALSE}
knitr::include_graphics("detection.pdf")
```

**Figure 3** Illustration of the frequency of selected Omicron mutations in wastewater samples. The x axis represents the number number of reads in a sample where that frequency was observed (e.g. in 3 out of 200 reads in a particular from Guelph). The size of the bubble represents the number of reads containing that mutation in a particular sample. The colours are the official colours of Western, Guelph, and Waterloo, respectively, and represent the lab that sequenced the data (not the location of the sample - this data is currently unavailable to us). The squares represent mutations that were not present in the samples \emph{OR} occurred in a location where the sample had zero coverage, so it would not be possible to tell whether these mutations were not present or merely not sequenced. The size of the square represents the log of the coverage (smallest squares have lowest coverage).


\clearpage
# Concluding Remarks

- We have identified mutations that are common in Omicron but rare in all other known variants, and (perhaps overenthusiastically) labelled these as "unique" mutations. 
  - Note that these mutations are likely present in unsequenced genomes; there is a lot of unsampled diversity out there. 
  - Even if a mutation is completely unique to Omicron compared to all known sequences, this does not imply that it's presence is a guarantee that we've observed Omicron.
  - Even if there existed a mutation that was completely unique to Omicron and not even present in any of the unsampled genomes, our observation of that mutation is subject to sequencing error.
- These "unique" mutations appear in a small number of our samples, and in very small numbers.
  - This is evidence that Omicron may present, but it is far from being conclusive proof.
- To make conclusions from our data, we would like to observe multiple "unique" mutations in one sample, and have this confirmed by another lab sequencing the same sample. 

<!-- Guelph Sample (naming convention?) -->
<!-- ========== -->


<!-- {data-width=250} -->
<!-- -------------------------------------------------- -->

<!-- ### Data Information -->

<!-- ```{r, echo=FALSE} -->
<!-- as.table(matrix(c("Collection Date", "SOme date", "collection location", "some location"), ncol = 2, byrow = T)) -->
<!-- ``` -->




<!-- -------------------------------------------------- -->

<!-- ### Reads Information -->



<!-- ![](sequence_reads_image.png) -->


<!-- ### Sequence Reads Analyzed -->
<!-- ```{r, echo=FALSE} -->
<!-- valueBox(400,icon = "ion-arrow-graph-up-right") -->
<!-- ``` -->


<!-- ### Mean Depth of Coverage -->

<!-- ```{r, echo=FALSE} -->
<!-- valueBox(500, icon = "ion-stats-bars") -->
<!-- ``` -->



<!-- Data Analysis -->
<!-- -------------------------------------------------- -->

<!-- ### Variants of Concern -->


<!-- Here we can add additional information currently I have variants of concern that have been found in a bar chart in their quantity -->

<!-- ```{r, echo=FALSE} -->
<!-- variants <- c("B.1.1.7","B.1.1.59", "A.1", "A.1", "A.1", "B.1.1.7") -->
<!-- variants <- data.frame(table(variants)) -->
<!-- ggplot(variants, aes(variants, Freq)) + geom_col() -->
<!-- ``` -->


