---
title: "Southern Ontario Wastewater Analysis Report"
author: "Erin Brintnell, Devan Becker, Gopi Gugan and Art Poon"
date: "`r Sys.Date()`"
always_allow_html: true
output:
  tufte::tufte_handout: default
  pdf_document:
    toc: false
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
---

```{r setup, include=FALSE}
library("here") # All paths relative to the directory with a .git folder
    # here("scripts", "plots.R") will always find plots.R,
    # no matter where this rmd's working directory is
library(flexdashboard)
library(dplyr)
library(tidyr)
library(knitr)
library(ggplot2)
library(kableExtra)
library(formattable)
library(here)

omicron <- read.csv(here("data", "omicron-mutations.csv"))

# load mutations for all lineages
background <- read.csv(here("data", "compare-out.csv"))

# what is the highest frequency that a mutation is found in ANY other lineage?
omicron$back.max <- apply(background[3:ncol(background)], 2, max)
omicron$n.lineages <- apply(background[3:ncol(background)], 2, function(y) sum(y>0))
omicron$which.max <- apply(background[3:ncol(background)], 2, 
                           function(y) background$lineage[which.max(y)])
omicron$back.mean <- apply(background[3:ncol(background)], 2, 
                           function(y) mean(y[y>0]))

# "omicron specific" mutations
specific <- omicron[omicron$freq >= 0.95 & omicron$back.max <= 0.05, ]
specific$label <- ifelse(specific$mutation=="None", paste(specific$type, specific$pos, specific$alt, sep="|"), specific$mutation)

# coverage
cvr <- read.csv(here("results", "cvr.csv"), row.names=1, stringsAsFactors = FALSE)
row.names(cvr) <- gsub("\\..+$", "", row.names(cvr))

# mutation counts
counts <- read.csv(here("results", "counts.csv"), row.names=1, stringsAsFactors = FALSE)
row.names(counts) <- gsub("\\..+$", "", row.names(counts))

```


<!-- ## Sourcing existing lineages -->

<!-- ```{r sourceascript, eval=FALSE} -->
<!-- source(here("scripts", "plot.R")) # won't run - files don't exist on my machine -->
<!-- ``` -->

# Background

* On November 23, 2021, a request for a new lineage designation was submitted to the CoV-lineages PANGO designation GitHub site, describing a new sub-lineage of B.1.1 sampled in Botswana, Hong Kong and South Africa with an excessive number of mutations.

* The request was reviewed and accepted on November 24, designating the new lineage as B.1.1.529.

* On November 26, the World Health Organization designated lineage B.1.1.529 as a variant of concern with the name "Omicron".

* On November 27, the CoV-lineages team uploaded a `constellation' file, describing a set of mutations for classifying genomes into the newly defined lineage.


# Deriving set of Omicron-specific mutations

We obtained 77 Omicron genomes from the GISAID database (obtained 2021-11-26). These data were processed using a script in our CoVizu (coronavirus visualization) system to generate a list of mutations relative to the reference genome, including indels, and uncalled bases for each genome.

Next, we screened about 3.4 million high-quality genomes for over 1,300 PANGO lineages from the GISAID database (obtained 2021-11-25) and calculated the frequencies of all $n=111$ Omicron mutations.
Mutations present in >95\% of Omicron genomes and <5\% of all other Pango lineages were selected for further downstream analysis of wastewater data. 

There were `r nrow(omicron)` mutations present in the Omicron genomes (n=77). Many mutations were present in other PANGO lineages (**Figure 1**). Nonetheless, `r nrow(specific)` mutations were present in >95\% of Omicron genomes and <5\% of all other Pango lineages (**Table 1**) which were used for downstream analyses.


```{r, echo = FALSE,fig.height=4.5, eval=FALSE}
ggplot(data = omicron, 
      mapping = aes(x = freq, y = back.max, 
        label = mutation, colour = downstream)) + 
    geom_point() + 
    labs(x = "Frequency in Omicron", y = "Frequency in Pango Lineages",
      colour = "UsedIn Downstream Analysis") +
    scale_y_log10() + 
    geom_jitter() + 
    theme(legend.position="bottom", 
      panel.background = element_blank(),  
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_line(colour = "grey"), 
      axis.line = element_line(colour = "black"), 
      text = element_text(size = 8)) +
    scale_color_manual(values=c("#999999", "#E69F00")) +
    annotate("line", x = c(0.95, 0.95), y = c(0, 1), 
      col = "#E69F00", linetype = 2) +
    annotate("line", x = c(0, 1), y = c(0.05, 0.05), 
      col = "#E69F00", linetype = 2)

```

```{r echo=FALSE, fig.height=4.5}
par(family='Palatino')
plot(omicron$freq, omicron$back.max, pch=21, col='white', 
     bg=ifelse(omicron$freq >= 0.95 & omicron$back.max <= 0.05, 'dodgerblue', 'firebrick'), xlab="Frequency in Omicron (n=77)", ylab="Frequency in all other lineages")
abline(a=0, b=1)
```
**Figure 1** Illustration of `r nrow(omicron)` mutations found throughout 77 Omnicron genomes. Mutations used in downstream analysis (blue) were present in >95\% of Omicron genomes and <5\% of all other Pango lineages (n=16). These numbers were chosen arbitrarily and will be explored in further studies.



\pagebreak
**Table 1** Mutations selected for downstream analysis in wastewater samples.  *Freq. Omicron* = observed frequency of mutation in *n*=77 Omicron genomes; *Max.bg.freq.* = maximum frequency of mutation in background (non-Omicron) genomes; *Mean.bg.freq.* = mean frequency of mutation in background genomes for all lineages where mutation was observed at least once; *# other lineages* = number of non-Omicron lineages in which mutation was observed at least once; *Most common* = PANGO nomenclature for lineage in which the mutation was observed at the highest relative frequency.
```{r echo=FALSE}
kable(specific[c('label', 'freq', 'back.max', 'back.mean', 'n.lineages',  'which.max')], digits=5,
      col.names=c("Mutation", "Freq. Omicron", "Max. bg. freq.", "Mean bg. freq.", "# other lineages", "Most common"), booktabs=T) %>%
  kable_paper("basic", full_width = F) %>%
    kable_styling(font_size = 9, latex_options = c("repeat_header"))
```


```{r, echo=FALSE, eval=FALSE}
omicron_v_pango %>%
  filter(downstream) %>% 
  select(mutation, freq, pango_freq) %>%
  arrange(pango_freq) %>% 
  kable(col.names = c("Mutation", "Frequency in Omicron", "Frequency in Pango lineages")) %>%
  column_spec(1, width = "18em")
```



# Screening wastewater samples

To date we have processed the raw sequence (FASTQ) data for `r nrow(cvr)` samples from laboratories based in Guelph, Waterloo and Western.
Briefly, we screened the sequence data for Illumina adapter sequence contamination and trimmed low quality bases, before mapping the reads to the SARS-CoV-2 reference genome using the program *minimap2*.
We captured the output stream of this program and parsed it using Python script adapted from our *CoVizu* pipeline.
This pipeline generates two CSV output files for each sample: (1) a CSV summarizing the number of reads mapped to each position in the reference genome (*i.e.*, coverage), and; (2) a CSV summarizing the frequencies of all mutations detected in the sample, including insertions and deletions.

We screened all mutation frequency outputs for entries that matched any of the 16 mutations that we previously determined to be diagnostic of Omicron genomes.
See Appendix for complete tables of mutation counts and coverage.
Positions without coverage with coded as missing data.
For each sample, we fit a binomial regression to the observed mutation counts and site-specific coverage to estimate the frequency of Omicron by maximum likelihood.
These estimates should be interpreted in the context of the expected sequencing error rate (see Appendix for an example of a run-specific distribution of error rates, based on empirical measurements for reads mapping to the $\phi$X174 control).
In further work, we will map these estimates directly to interpreting our binomial regressions.
Otherwise, we will assume that the overall error rate is about 1%.


```{r echo=FALSE, warning=FALSE, fig.height=6, fig.width=4}

probs <- c()
lo <- c()
hi <- c()
for (i in 1:nrow(counts)) {
  y <- as.integer(counts[i, 1:16])  # number of "successes"
  n <- as.integer(cvr[i, 1:16])  # number of trials
  p <- tryCatch({
    fit <- glm(cbind(y, n) ~ 1, family='binomial')
    exp(fit$coef) / (1+exp(fit$coef))  # probability
    }, 
    error = function(cond) { 
      return (NA) 
    })
  
  probs <- c(probs, p)
  if (is.na(p)) {
    lo <- c(lo, NA)
    hi <- c(hi, NA)
  } else {
    suppressMessages(ci <- confint(fit))
    lo <- c(lo, exp(ci[1]) / (1+exp(ci[1])))
    hi <- c(hi, exp(ci[2]) / (1+exp(ci[2])))
  }
}
```



```{r echo=FALSE, warning=FALSE, fig.height=6, fig.width=4}
idx <- which(counts$lab=='guelph')
barplot(as.numeric(probs)[idx], horiz=T, main="Guelph", adj=0, cex.main=1, 
        names.arg=gsub("\\.[a-z0-9]+$", "", counts$sample[idx]), 
        las=1, cex.names=0.5, xlim=c(0, 0.05), cex.axis=0.7,
        xlab="Estimated frequency", cex.lab=0.9)
segments(x0=lo[idx], x1=hi[idx], y0=(1:length(idx)-0.45)*1.2, lwd=2)
abline(v=0.01, lty=2)
```
**Figure 2** Barplot summarizing the estimated frequencies of Omicron for samples processed by the Guelph laboratory.  Line segments correspond to the 95% confidence intervals.  A dashed line indicates a hypothetical sequencing error rate of 1%.


```{r echo=FALSE, warning=FALSE, fig.height=6, fig.width=4}
idx <- which(counts$lab=='waterloo')
barplot(as.numeric(probs)[idx], horiz=T, main="Waterloo", adj=0, cex.main=1, 
        names.arg=gsub("\\.[a-z0-9]+$", "", counts$sample[idx]), 
        las=1, cex.names=0.5, xlim=c(0, 0.05), cex.axis=0.7,
        xlab="Estimated frequency", cex.lab=0.9)
segments(x0=lo[idx], x1=hi[idx], y0=(1:length(idx)-0.45)*1.2, lwd=2)
abline(v=0.01, lty=2)
```
**Figure 3** Barplot summarizing the estimated frequencies of Omicron for samples processed by the Waterloo laboratory.  Line segments correspond to the 95% confidence intervals.  A dashed line indicates a hypothetical sequencing error rate of 1%.


As of time of writing, there were no sample data sets available from Western University with sufficient coverage to carry out this analysis.


# Conclusions

As of this time, there is no evidence that any genomic material derived from copies of SARS-CoV-2 variant of concern Omicron were present in samples for which we have sufficient coverage.
Although we detected mutations associated with Omicron in the sequence data, none of these mutations were found at a frequency sufficiently high to reject the alternative explanation that the mutations were the result of sequencing error, using a conservative threshold of 1% for the sequencing platform used by the labs.



\clearpage

# Appendix

## Error rates

These plots were generated from a single run of an Illumina sequencing platform.  The data were extracted from a binary InterOp file using a Python script.  The left and right panels correspond to the forward and reverse reads, respectively.  Each point represents a tile-cycle combination.  In this particular case, the run was configured to run for 150 cycles in either direction.  The number of tiles varies with the specific model of sequencing platform.  We note that a small number of tile-cycle combinations are associated with unusually high error rates, which is a known issue.  For amplicon-based sequencing applications, these outliers can be associated with specific sites of the genome.

```{r echo=FALSE, fig.height=5, fig.width=9, fig.fullwidth=TRUE}
erates <- read.csv(here("reports", "error-rates.csv"))
par(mfrow=c(1,2), family="Palatino")
plot(erates$cycle[erates$cycle>0], erates$errorrate[erates$cycle>0], 
     bg=rgb(1,0,0,0.25), cex=1, pch=21, lwd=0.75, main="First read",
     xlab="Cycle number", ylab="Empirical error rate (%)", 
     ylim=c(0, max(erates$errorrate, na.rm=T)))
plot(-erates$cycle[erates$cycle<0], erates$errorrate[erates$cycle<0],
     bg=rgb(0,0,1,0.25), cex=1, pch=21, lwd=0.75, main="Second read",
     xlab="Cycle number", ylab="Empirical error rate (%)",
     ylim=c(0, max(erates$errorrate, na.rm=T)))
```


## Coverage

```{r, echo=FALSE, warning=FALSE}
cvr[1:ncol(cvr)] <- lapply (cvr[1:ncol(cvr)], function(x){
  cell_spec(x, color = "black", background = spec_color(x, option = "A", begin = 0.2,  direction = -1 , scale_from = c(0,10000)))
})
if (any(grepl("/", row.names(cvr)))) {
  row.names(cvr) <- sapply(strsplit(row.names(cvr), split = "/"), function(x) {
    strsplit(x[length(x)], split = "\\.")[[1]][1]
  })
}

 kbl(cvr, escape = F, align = "c", longtable = TRUE) %>%
  kable_paper("basic", full_width = F) %>%
    kable_styling(font_size = 7, latex_options = c("repeat_header")) %>% column_spec(2:(ncol(cvr)+1), width = "0.62cm") %>%
      column_spec(1, width = "1cm") %>% row_spec(0,angle = 90)
```


\clearpage

## Mutation counts

```{r, echo=FALSE, warning=FALSE}
if (any(grepl("/", row.names(counts)))) {
  row.names(counts) <- sapply(strsplit(row.names(counts), split = "/"), function(x) {
    strsplit(x[length(x)], split = "\\.")[[1]][1]
  })
}
col_nams <- colnames(counts)
row_nams <- rownames(counts)
counts <- data.frame(matrix(as.numeric(unlist(counts)), nrow = nrow(counts)))
colnames(counts) <- col_nams
counts <- counts[,-c(20,19,18, 17)]
rownames(counts) <- row_nams

counts[1:ncol(counts)] <- lapply (counts[1:ncol(counts)], function(x){
  cell_spec(x, color = "black", background = spec_color(x, option = "A", begin = 0.5, direction = -1, na_color = "white", scale_from = c(0,50)))
})

 kbl(counts, escape = F, align = "c", longtable = TRUE) %>%
  kable_paper("basic", full_width = F) %>%
    kable_styling(font_size = 7, latex_options = c("repeat_header")) %>% column_spec(2:(ncol(counts)+1), width = "0.62cm") %>%
      column_spec(1, width = "1cm") %>% row_spec(0,angle = 90)
```


